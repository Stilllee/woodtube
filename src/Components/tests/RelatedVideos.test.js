import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { withAllContexts, withRouter } from "../../tests/utils";

import React from "react";
import RelatedVideos from "../RelatedVideos";
import { Route } from "react-router-dom";
import { fakeVideos } from "../../tests/videos";

describe("RelatedVideos", () => {
  const fakeYoutube = {
    relatedVideos: jest.fn(),
  };

  afterEach(() => fakeYoutube.relatedVideos.mockReset());

  it("올바르게 렌더링", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    const { asFragment } = renderRelatedVideos();

    await waitForElementToBeRemoved(screen.getByText("Loading..."));
    expect(asFragment()).toMatchSnapshot();
  });

  it("연관 비디오를 올바르게 렌더링", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith("id");
    await waitFor(() => {
      expect(screen.getAllByRole("listitem")).toHaveLength(fakeVideos.length);
    });
  });

  it("로딩 렌더링", () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("에러 렌더링", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => {
      throw new Error("error");
    });

    renderRelatedVideos();
    await waitFor(() => {
      expect(screen.getByText("Something is wrong 😖")).toBeInTheDocument();
    });
  });

  function renderRelatedVideos() {
    return render(
      withAllContexts(
        withRouter(<Route path="/" element={<RelatedVideos id="id" />} />),
        fakeYoutube
      )
    );
  }
});
