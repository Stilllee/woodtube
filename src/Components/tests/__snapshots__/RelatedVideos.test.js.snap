// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos 올바르게 렌더링 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col gap-4"
  >
    <li
      class="flex gap-1 cursor-pointer hover:scale-105 transition-transform rounded-xl"
    >
      <img
        alt="제목"
        class="w-60 mr-2 rounded-xl"
        src="http://example.com/image.jpg"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
          title="제목"
        >
          제목
        </p>
        <p
          class="text-sm opacity-80"
        >
          채널명
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
    <li
      class="flex gap-1 cursor-pointer hover:scale-105 transition-transform rounded-xl"
    >
      <img
        alt="제목2"
        class="w-60 mr-2 rounded-xl"
        src="http://example.com/image2.jpg"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
          title="제목2"
        >
          제목2
        </p>
        <p
          class="text-sm opacity-80"
        >
          채널명2
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
